base/math.jl

```diff
@@ -537,54 +543,56 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
+hypot(x::Number, y::Number; kwargs...) = hypot(promote(x, y)...; kwargs...)
+hypot(x::Complex, y::Complex; kwargs...) = hypot(promote(abs(x),abs(y))...; kwargs...)
+hypot(x::Integer, y::Integer; kwargs...) = hypot(promote(float(x), float(y))...; kwargs...)
+function hypot(x::T,y::T; correctlyrounded::Bool=false) where T<:AbstractFloat
+    #Return Inf if either or both imputs is Inf (Compliance with IEEE754)
+    if isinf(x) || isinf(y)
+        return convert(T,Inf)
     end
-    if iszero(ax)
-        r = ay / oneunit(ax)
-    else
-        r = ay / ax
+
+    # Order the operands
+    ax,ay = abs(x), abs(y)
+    if ay > ax
+        ax,ay = ay,ax
     end
 
-    rr = ax * sqrt(1 + r * r)
+    # Widely varying operands
+    if ay <= ax*sqrt(eps(T)/2)  #Note: This also gets ay == 0
+        return ax
+    end
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
-    if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
+    # Operands do not vary widely
+    scale = eps(sqrt(floatmin(T)))  #Rescaling constant
+    if ax > sqrt(floatmax(T)/2)
+        ax = ax*scale
+        ay = ay*scale
+        scale = inv(scale)
+    elseif ay < sqrt(floatmin(T))
+        ax = ax/scale
+        ay = ay/scale
     else
-        return rr
+        scale = one(scale)
     end
+    h = sqrt(muladd(ax,ax,ay*ay))
+    if Base.Math.FMA_NATIVE | correctlyrounded
+        h = sqrt(fma(ax,ax,ay*ay))
+        h_sq = h*h
+        ax_sq = ax*ax
+        h -= (fma(-ay,ay,h_sq-ax_sq) + fma(h,h,-h_sq) - fma(ax,ax,-ax_sq))/(2*h)
+    else
+        if h <= 2*ay
+            delta = h-ay
+            h -= muladd(delta,delta-2*(ax-ay),ax*(2*delta - ax))/(2*h)
+        else
+            delta = h-ax
+            h -= muladd(delta,delta,muladd(ay,(4*delta-ay),2*delta*(ax-2*ay)))/(2*h)
+        end
+    end
+    h*scale
 end
 
-"""
-    hypot(x...)
-
-Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
-"""
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
-
-atan(y::Real, x::Real) = atan(promote(float(y),float(x))...)
```

Comment:
You probably didn't mean to delete all these methods of other functions, right?

Reply:
also line 550 typo "imputs"

Reply:
No I didn't. I'm brand new to github and I make some pretty boneheaded mistakes.

Reply:
I'll give it a try.